File: algorithms/python/jumpGameV/jumpGameV.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/jump-game-v/
# Author: Miao Zhang
# Date:   2021-04-25

class Solution:
    def maxJumps(self, arr: List[int], d: int) -> int:
        n = len(arr)
        memo = [[] for _ in range(n)]
        for i in range(len(arr)):
            memo[i].append(arr[i])
            memo[i].append(i)
        memo.sort()
        dp = [1 for _ in range(n)]
        for v, i in memo:
            j = i + 1
            while j < min(n, i + d + 1) and arr[i] > arr[j]:
                dp[i] = max(dp[i], dp[j] + 1)
                j += 1
            j = i - 1
            while j > max(-1, i - d - 1) and arr[i] > arr[j]:
                dp[i] = max(dp[i], dp[j] + 1)
                j -= 1
        return max(dp)
            

File: algorithms/cpp/jumpGameV/jumpGameV.cpp
// Source: https://leetcode.com/problems/jump-game-v/
// Author: Miao Zhang
// Date:   2021-04-25

/********************************************
 * dp[i] = 1 or max(dp[j]) + 1 if i can jump to j
********************************************/
class Solution {
public:
    int maxJumps(vector<int>& arr, int d) {
        int n = arr.size();
        vector<int> m(n);
        function<int(int)> dp = [&] (int i) {
            if (m[i]) return m[i];
            int res = 1;
            for (int j = i + 1; j < min(n, i + d + 1) && arr[i] > arr[j]; j++)
                res = max(res, dp(j) + 1);
            for (int j = i - 1; j >= max(0, i - d) && arr[i] > arr[j]; j--)
                res = max(res, dp(j) + 1);
            return m[i] = res;
        };
        int res = 0;
        for (int i = 0; i < arr.size(); i++) {
            res = max(res, dp(i));
        }
        return res;
    }
};
